src/pkg/artifact/dao/dao.go

```diff
@@ -0,0 +1,114 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dao
+
+import (
+	"context"
+	"github.com/goharbor/harbor/src/internal/orm"
+	"github.com/goharbor/harbor/src/pkg/q"
+)
+
+// DAO is the data access object interface for artifact
+type DAO interface {
+	// Count returns the total count of artifacts according to the query
+	Count(ctx context.Context, query *q.Query) (total int64, err error)
+	// List artifacts according to the query
+	List(ctx context.Context, query *q.Query) (artifacts []*Artifact, err error)
+	// Get the artifact specified by ID
+	Get(ctx context.Context, id int64) (*Artifact, error)
+	// Create the artifact
+	Create(ctx context.Context, artifact *Artifact) (id int64, err error)
+	// Delete the artifact specified by ID
+	Delete(ctx context.Context, id int64) (err error)
+	// Update updates the artifact. Only the properties specified by "props" will be updated if it is set
+	Update(ctx context.Context, artifact *Artifact, props ...string) (err error)
+	// CreateReference creates the artifact reference
+	CreateReference(ctx context.Context, reference *ArtifactReference) (id int64, err error)
+	// ListReferences lists the artifact references according to the query
+	ListReferences(ctx context.Context, query *q.Query) (references []*ArtifactReference, err error)
+	// DeleteReferences deletes the references referenced by the artifact specified by parent ID
+	DeleteReferences(ctx context.Context, parentID int64) (err error)
+}
+
+// New returns an instance of the default DAO
+func New() DAO {
+	return &dao{}
+}
+
+type dao struct{}
+
+func (d *dao) Count(ctx context.Context, query *q.Query) (int64, error) {
+	return orm.GenerateQuerySetter(ctx, &Artifact{}, query, false).Count()
+}
+func (d *dao) List(ctx context.Context, query *q.Query) ([]*Artifact, error) {
+	artifacts := []*Artifact{}
+	if _, err := orm.GenerateQuerySetter(ctx, &Artifact{}, query, true).All(&artifacts); err != nil {
+		return nil, err
+	}
+	return artifacts, nil
+}
+func (d *dao) Get(ctx context.Context, id int64) (*Artifact, error) {
+	artifact := &Artifact{
+		ID: id,
+	}
+	if err := orm.GetOrmer(ctx).Read(artifact); err != nil {
+		if e, ok := orm.IsNotFoundError(err, "artifact %d not found", id); ok {
+			err = e
+		}
+		return nil, err
+	}
+	return artifact, nil
+}
+func (d *dao) Create(ctx context.Context, artifact *Artifact) (int64, error) {
+	id, err := orm.GetOrmer(ctx).Insert(artifact)
+	if e, ok := orm.IsConflictError(err, "artifact %s already exists under the repository %d",
+		artifact.Digest, artifact.RepositoryID); ok {
+		err = e
+	}
+	return id, err
+}
+func (d *dao) Delete(ctx context.Context, id int64) error {
+	_, err := orm.GetOrmer(ctx).Delete(&Artifact{
+		ID: id,
+	})
+	return err
+}
+func (d *dao) Update(ctx context.Context, artifact *Artifact, props ...string) error {
+	_, err := orm.GetOrmer(ctx).Update(artifact, props...)
```

Comment:
update() can return orm.ErrNoRows.
        _, err := orm.GetOrmer(ctx).Update(artifact, props...)
	if err != nil {
		if errors.Is(err, orm.ErrNoRows) {
			return id, internal_errors.New(err).WithCode(internal_errors.NotFoundCode).
				WithMessage(fmt.Sprintf("the artifact %d is not found.", id))
		}
		return id, err
	}
	return id, nil

Reply:
I don't think DB will return errors if artifact does not exist.  But I DO think we should not ignore the first return value of Update.

Reply:
When affected rows is 0, return not found error